Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,294 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PlatformSpeechSynthesizer.h"
+
+#if ENABLE(SPEECH_SYNTHESIS) && USE(SPIEL)
+
+#include "GStreamerCommon.h"
+#include "PlatformSpeechSynthesisUtterance.h"
+#include "PlatformSpeechSynthesisVoice.h"
+#include "WebKitAudioSinkGStreamer.h"
+#include <spiel/spiel.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/text/MakeString.h>
+
+namespace WebCore {
+
+#define URI_PREFIX "urn:webkit-tts:spiel:"_s
+
+class SpielSpeechWrapper {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(SpielSpeechWrapper);
+    WTF_MAKE_NONCOPYABLE(SpielSpeechWrapper);
+public:
+    explicit SpielSpeechWrapper(const PlatformSpeechSynthesizer&, Function<void()>&&);
+    ~SpielSpeechWrapper();
+
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> initializeVoiceList();
+    void pause();
+    void resume();
+    void speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&&);
+    void cancel();
+    void clearUtterance();
+
+private:
+    void finishSpeakerInitialization();
+    String generateVoiceURI(const GRefPtr<SpielVoice>&, const String& language);
+
+    const PlatformSpeechSynthesizer& m_platformSynthesizer;
+    Function<void()> m_speakerCreatedCallback;
+    GRefPtr<GstElement> m_sink;
+    GRefPtr<SpielSpeaker> m_speaker;
+    HashMap<String, GRefPtr<SpielVoice>> m_voices;
+    RefPtr<PlatformSpeechSynthesisUtterance> m_utterance;
+};
+
+SpielSpeechWrapper::SpielSpeechWrapper(const PlatformSpeechSynthesizer& synthesizer, Function<void()>&& speakerCreatedCallback)
+    : m_platformSynthesizer(synthesizer)
+    , m_speakerCreatedCallback(WTFMove(speakerCreatedCallback))
+{
+    ensureGStreamerInitialized();
+    registerWebKitGStreamerElements();
+
+    m_sink = createPlatformAudioSink("speech"_s);
+    if (!m_sink) {
+        GST_ERROR("Failed to create GStreamer audio sink element");
+        return;
+    }
+
+    spiel_speaker_new(nullptr, [](GObject*, GAsyncResult* result, gpointer userData) {
```

Comment:
I'm curious of why you use the async API instead of the sync one.

Reply:
sync dbus calls should be avoided when possible. Our main loop is busy enough already.